### PR TITLE
Run shellcheck as a github workflow

### PR DESCRIPTION
**Pull request by hulthe**
`shellcheck` lints scripts for common footguns. This PR adds `shellcheck` as a GitHub action, and also attempts to fix all the warnings we currently have.

Scripts are finnicky, and `shellcheck` has a tendency to give false positives, so the changed scripts should be tested before merging if the changes are non-trivial. 

- - -
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876)

### CODE REVIEW

File: ci/ios/buildserver-build-ios.sh [hunk 1]
```diff
@@ -2,7 +2,7 @@
 set -eu
 shopt -s nullglob
 
-TAG_PATTERN_TO_BUILD=("^ios/")
+TAG_PATTERN_TO_BUILD="^ios/"
```
Comment (hulthe) [inline ci/ios/buildserver-build-ios.sh:5]: https://www.shellcheck.net/wiki/SC2128

File: ci/ios/buildserver-build-ios.sh [hunk 2]
```diff
@@ -93,7 +93,7 @@ function run_build_loop() {
 
         run_git fetch --prune --tags 2> /dev/null || continue
         local tags
-        tags=( $(run_git tag | grep "$TAG_PATTERN_TO_BUILD") )
+        mapfile -t tags < <(run_git tag | grep "$TAG_PATTERN_TO_BUILD")
```
Comment (hulthe) [inline ci/ios/buildserver-build-ios.sh:96]: https://www.shellcheck.net/wiki/SC2207

Comment (hulthe) [reply ci/ios/buildserver-build-ios.sh:96]: note: only supported on bash 4.4+, but that was released in 2017 so should be fine.
also not posix compatible, but I assume that's not something we care about

File: ci/ios/run-in-vm.sh [hunk 3]
```diff
@@ -24,7 +24,7 @@ sleep 10
 
 # apparently, there's a difference between piping into zsh like this and doing
 # a <(echo $SCRIPT).
-cat "$SCRIPT" | ssh admin@"$(tart ip "$VM_NAME")" bash /dev/stdin
+ssh admin@"$(tart ip "$VM_NAME")" bash /dev/stdin < "$SCRIPT"
```
Comment (hulthe) [inline ci/ios/run-in-vm.sh:27]: https://www.shellcheck.net/wiki/SC2002

Comment (hulthe) [reply ci/ios/run-in-vm.sh:27]: not sure what the comment above is about? should be looked into before merging

File: dist-assets/linux/before-remove.sh [hunk 4]
```diff
@@ -10,7 +10,7 @@ pkill -9 -x "mullvad-gui" || true
 is_number_re='^[0-9]+$'
 # Check if we're running during an upgrade step on Fedora
 # https://fedoraproject.org/wiki/Packaging:Scriptlets#Syntax
-if [[ "$1" =~ $is_number_re ]] && [ $1 -gt 0 ]; then
+if [[ "$1" =~ $is_number_re ]] && [ "$1" -gt 0 ]; then
```
Comment (hulthe) [inline dist-assets/linux/before-remove.sh:13]: https://www.shellcheck.net/wiki/SC2086

File: dist-assets/linux/mullvad-gui-launcher.sh [hunk 5]
```diff
@@ -13,10 +13,10 @@ if [ "${XDG_SESSION_TYPE:-""}"  = "wayland" ] && \
...
-    WAYLAND_FLAGS="--ozone-platform=wayland --enable-features=WaylandWindowDecorations"
+    WAYLAND_FLAGS=( "--ozone-platform=wayland" "--enable-features=WaylandWindowDecorations" )
 else
-    WAYLAND_FLAGS=""
+    WAYLAND_FLAGS=()
 fi
 
 SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
-exec "$SCRIPT_DIR/mullvad-gui" $SANDBOX_FLAG $WAYLAND_FLAGS "$@"
+exec "$SCRIPT_DIR/mullvad-gui" "$SANDBOX_FLAG" "${WAYLAND_FLAGS[@]}" "$@"
```
Comment (hulthe) [inline dist-assets/linux/mullvad-gui-launcher.sh:22]: https://www.shellcheck.net/wiki/SC2086

File: dist-assets/pkg-scripts/postinstall [hunk 6]
```diff
@@ -5,7 +5,7 @@ set -eu
 LOG_DIR=/var/log/mullvad-vpn
 
 mkdir -p $LOG_DIR
-exec 2>&1 > $LOG_DIR/postinstall.log
+exec > $LOG_DIR/postinstall.log 2>&1
```
Comment (hulthe) [inline dist-assets/pkg-scripts/postinstall:8]: https://www.shellcheck.net/wiki/SC2069
I assume this wasn't intentional

File: dist-assets/pkg-scripts/preinstall [hunk 7]
```diff
@@ -7,7 +7,7 @@ INSTALL_DIR=$2
 
 mkdir -p $LOG_DIR
 chmod 755 $LOG_DIR
-exec 2>&1 > $LOG_DIR/preinstall.log
+exec > $LOG_DIR/preinstall.log 2>&1
```
Comment (hulthe) [inline dist-assets/pkg-scripts/preinstall:10]: https://www.shellcheck.net/wiki/SC2069
I assume this wasn't intentional

File: dist-assets/uninstall_macos.sh [hunk 8]
```diff
@@ -52,7 +52,7 @@ if [[ "$REPLY" =~ [Yy]$ ]]; then
     done
 fi
 
-read -p "Do you want to delete the Mullvad VPN settings? (y/n) "
+read -r -p "Do you want to delete the Mullvad VPN settings? (y/n) "
```
Comment (hulthe) [inline dist-assets/uninstall_macos.sh:55]: https://www.shellcheck.net/wiki/SC2162

File: ios/build-wireguard-go.sh [hunk 9]
```diff
@@ -38,4 +38,4 @@ WIREGUARD_KIT_GO_PATH="$RESOLVED_SOURCE_PACKAGES_PATH/checkouts/wireguard-apple/
 echo "WireGuardKitGo path resolved to $WIREGUARD_KIT_GO_PATH"
 
 # Run make
-/usr/bin/make -C "$WIREGUARD_KIT_GO_PATH" $ACTION
\ No newline at end of file
+/usr/bin/make -C "$WIREGUARD_KIT_GO_PATH" "$ACTION"
```
Comment (hulthe) [inline ios/build-wireguard-go.sh:41]: https://www.shellcheck.net/wiki/SC2086

File: test/ci-runtests.sh [hunk 10]
```diff
@@ -196,7 +196,7 @@ function run_tests_for_os {
 
     RUST_LOG=debug cargo run --bin test-manager \
         run-tests \
-        --account "${ACCOUNT_TOKEN}" \
+        --account "${ACCOUNT_TOKEN:?Error: ACCOUNT_TOKEN not set}" \
```
Comment (hulthe) [inline test/ci-runtests.sh:199]: https://www.shellcheck.net/wiki/SC2153

File: wireguard/libwg/build-android.sh [hunk 11]
```diff
@@ -41,7 +41,7 @@ for arch in ${ARCHITECTURES:-armv7 aarch64 x86_64 i686}; do
     esac
 
     # Build Wireguard-Go
-    echo "$(pwd)"
+    pwd
```
Comment (hulthe) [inline wireguard/libwg/build-android.sh:44]: https://www.shellcheck.net/wiki/SC2005

File: wireguard/libwg/build-android.sh [hunk 12]
```diff
@@ -51,7 +51,8 @@ for arch in ${ARCHITECTURES:-armv7 aarch64 x86_64 i686}; do
 
     # Create the directories with RWX permissions for all users so that the build server can clean
     # the directories afterwards
-    mkdir -m 777 -p "$(dirname "$STRIPPED_LIB_PATH")"
+    mkdir -p "$(dirname "$STRIPPED_LIB_PATH")"
+    chmod 777 "$(dirname "$STRIPPED_LIB_PATH")"
```
Comment (hulthe) [inline wireguard/libwg/build-android.sh:55]: https://www.shellcheck.net/wiki/SC2174

Comment (hulthe) [reply wireguard/libwg/build-android.sh:55]: note: I assume this is not what the author intended. Worth fixing?

**Review by faern — CHANGES_REQUESTED** [review]
Some of the scripts (most notably the iOS scripts) I don't feel like I can verify correctly. Can you please have someone in the iOS team test these changes out and verify they work as expected? I know we have previously had issuse where shellcheck recommendations break iOS scripts.

> Reviewed 7 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876#-:-Nromxo0B48Xj5VPcfQ8:btimi9s)* status: 7 of 12 files reviewed, 16 unresolved discussions (waiting on @hulthe)

___
*[`.github/workflows/shellcheck.yml` line 5 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876#-Nrohrd33WPvL1yP2MGT:-Nrohrd33WPvL1yP2MGU:b-el4mvw) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/668c58e9c783b76cf223d41a8f8069aff3976eeb/.github/workflows/shellcheck.yml#L5)):*
> ```YAML
>   push:
>     branches:
>       - main
> ```

Is this needed on push? We don't run other CI jobs on push really, except a few, but they have special reasons. As long as we check all PRs, `main` should never get corrupted. I'm not saying I'm completely against this, just that it's different from the pattern we usually have. I'm open to suggested improvements.
___
*[`.github/workflows/shellcheck.yml` line 8 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876#-NrompKXB0mMTPGYmvip:-NrompKXB0mMTPGYmviq:b-gvzj9l) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/668c58e9c783b76cf223d41a8f8069aff3976eeb/.github/workflows/shellcheck.yml#L8)):*
> ```YAML
> 
> name: "Trigger: Push action"
> permissions: {}
> ```

Are these two lines needed? we don't really specify them for other jobs. I think this will work fine without? :thinking: 
___
*[`.github/workflows/shellcheck.yml` line 17 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876#-NroigOd7FckwfNvBLLR:-NroigOd7FckwfNvBLLS:b-5oru94) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/668c58e9c783b76cf223d41a8f8069aff3976eeb/.github/workflows/shellcheck.yml#L17)):*
> ```YAML
>       - uses: actions/checkout@v3
>       - name: Run ShellCheck
>         uses: ludeeus/action-shellcheck@master
> ```

Would it make sense to pin this to a release of the action instead of `master`? Mostly to stop sudden breakage. The CI fails often enough without depending on moving targets :joy: Seems like `2.0.0` is the latest release.
___
*[`.github/workflows/shellcheck.yml` line 20 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876#-Nroj5_B3xFybziwf_8Y:-Nroj5_B3xFybziwf_8Z:b-l10nd4) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/668c58e9c783b76cf223d41a8f8069aff3976eeb/.github/workflows/shellcheck.yml#L20)):*
> ```YAML
>         with:
>           ignore_names: >-
>             gradlew
> ```

We really only have a single `gradlew` file that we want to ignore. So it feels like we can specify this more exactly with:
```
ignore_paths: >-
  android/gradlew
```
I feel like this more exactly express what we want to ignore.

**Review by faern — COMMENTED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876)* status: 8 of 12 files reviewed, 11 unresolved discussions

**Review by buggmagnet — CHANGES_REQUESTED** [review]
> Reviewed 10 of 12 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876)* status: all files reviewed, 14 unresolved discussions (waiting on @hulthe)

___
*[`ci/ios/buildserver-build-ios.sh` line 96 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876#-NsCXJL75CKLF9674Giw:-NsCXJL75CKLF9674Gix:b-3bm7fz) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/7d84ea9d86a6545615203e85ead131c8d9761c3b/ci/ios/buildserver-build-ios.sh#L96)):*
> ```Shell
>         run_git fetch --prune --tags 2> /dev/null || continue
>         local tags
>         mapfile -t tags < <(run_git tag | grep "$TAG_PATTERN_TO_BUILD")
> ```

```
bash-3.2$ mapfile 
bash: mapfile: command not found
```

It looks like this builtin does not exists on bash for macOS. I think we should revert this change.
___
*[`ci/ios/run-in-vm.sh` line 27 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876#-NsCXs5LEZwYHwuCPvji:-NsCXs5LEZwYHwuCPvjj:byfe3a6) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/7d84ea9d86a6545615203e85ead131c8d9761c3b/ci/ios/run-in-vm.sh#L27)):*
> ```Shell
> # apparently, there's a difference between piping into zsh like this and doing
> # a <(echo $SCRIPT).
> ssh admin@"$(tart ip "$VM_NAME")" bash /dev/stdin < "$SCRIPT"
> ```

We can remove the above comment as it will be invalid after this change.
I just tested this, and that seems to work, so this change should be fine. 
___
*[`ios/build-wireguard-go.sh` line 41 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876#-NsCa7mpD-KAt6adXSa0:-NsCa7mpD-KAt6adXSa1:bks8e26) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/7d84ea9d86a6545615203e85ead131c8d9761c3b/ios/build-wireguard-go.sh#L41)):*
> ```Shell
> 
> # Run make
> /usr/bin/make -C "$WIREGUARD_KIT_GO_PATH" "$ACTION"
> ```

This is intentionally done without quotes, this script is ran from within Xcode, and it won't always have an argument. 

If we add quotes here, we will pass an empty string as an argument which will make our build fail, please revert this change.

**Review by hulthe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876)* status: 9 of 12 files reviewed, 14 unresolved discussions (waiting on @buggmagnet)

___
*[`ci/ios/buildserver-build-ios.sh` line 96 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876#-NsCXJL75CKLF9674Giw:-NsD2Cyc1-C5IVqv4SZr:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/7d84ea9d86a6545615203e85ead131c8d9761c3b/ci/ios/buildserver-build-ios.sh#L96)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

```
bash-3.2$ mapfile 
bash: mapfile: command not found
```

It looks like this builtin does not exists on bash for macOS. I think we should revert this change.
</blockquote></details>

Done.
___
*[`ci/ios/run-in-vm.sh` line 27 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876#-NsCXs5LEZwYHwuCPvji:-NsD2EQV1ElrTbaLhbEc:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/7d84ea9d86a6545615203e85ead131c8d9761c3b/ci/ios/run-in-vm.sh#L27)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

We can remove the above comment as it will be invalid after this change.
I just tested this, and that seems to work, so this change should be fine.
</blockquote></details>

Done.
___
*[`ios/build-wireguard-go.sh` line 41 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876#-NsCa7mpD-KAt6adXSa0:-NsD2Ej06cFfzXBLJlT_:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/7d84ea9d86a6545615203e85ead131c8d9761c3b/ios/build-wireguard-go.sh#L41)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

This is intentionally done without quotes, this script is ran from within Xcode, and it won't always have an argument. 

If we add quotes here, we will pass an empty string as an argument which will make our build fail, please revert this change.
</blockquote></details>

Done.

**Review by buggmagnet — APPROVED** [review]
My review pertains only to iOS specific changes, but :lgtm:
Thank you for doing this!

> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876#-:-NsD2aSxEKZ_bQMcUt1D:b4w8dwd)* status: all files reviewed, 11 unresolved discussions

**Review by faern — APPROVED** [review]
> Reviewed 1 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5876)* status: all files reviewed, 11 unresolved discussions

**Pull request merged**